README.md
# Playfair Evolutionary Solver

Written in **Python 3.9**, but other versions may also work.

This is an evolutionary algorithm implementation of a Playfair cipher solver.
The grid is represented as a string of characters A-Z (excluding J) and will
iterate through generations combining solutions and mutating them until the
known plain text is discovered. This does not guarantee a correct solution, but
will eventually find a valid solution that can produce the plain text given.

A design decision not to use any additional libraries such as numpy was made
to make the program easier to run without having to set up a virtual environment
a language such as and install the libraries.

## Setup

There are no dependencies other than Python 3. Just clone the repo and get
cracking!

```
git clone https://github.com/GTBebbo/PlayfairEvolutionarySolver.git
```

## Usage

Place the cipher text in a `cipher.text` file in the same directory as
`main.py`.

Place the known plain text in a `plain.text` file in the same directory as
`main.py`.

To run the solver run:

```
python main.py
```

__If you need to change the Playfair algorithm itself, you can modify the
`ALPHABET` parameter in the `settings.py` file for example to change the
excluded `J` to another letter.__

## How to solve

This does not guarantee a solution, this is purely an evolutionary algorithm
to find a solution that fits or nearly fits the plaintext. It works best with
longer (50+ character) ciphers and plain texts. Smaller ciphers can produce
valid but incorrect tables too easily.

You may need to modify the settings in `settings.py` to achieve better results.

Often when using a small amount of plaintext, the a valid table may be found
that doesn't decrypt the entire ciphertext. This is an iterative process and you
should use the found solution to guess at some previously unknown plaintext.
By expanding the plaintext, you will have a better chance of finding the correct
solution.

For example, given the following:

```
# Ciphertext
YL NR EX RQ AT QB ML NG OV LO OI DY
# Plaintext
WOULD YOU LIKE TO
```

You may find a result like this:

```
WOULD YOU LIKE TO PLAYAGOMEOFNSEOLATYD
```

We can now guess at some of the previously unknown ciphertext and update our
plaintext file.

```
WOULD YOU LIKE TO PLAY A GAME
```

Running the algorithm again with this new plaintext gives us the correct result:

```
WOULD YOU LIKE TO PLAY A GAME OF MONOPOLY X
```

__This is the table to decrypt the above if you were interested:__

```
ALGOR
ITHMS
BCDEF
KNPQU
VWXYZ
```

## Furture Improvements

**Mutations**: Currently the mutation step swaps random letters in the table,
an improvement could swap entire rows and columns.

**Playfair Semantics**: The current algorithm ignores the sematics of Playfair,
for example the fact characters should form rectangles or lines.

**Language**: There is a possibility to speed the algorithm up by a lot by
re-implementing the program in a language such as C++.

**Automatic Plaintext Recognition**: By using a dictionary of words, we can
compare the decrypted plaintext result to try to automatically expand the known
plaintext to prevent the user from manually doing so.

**Plaintext Anywhere**: Currently the algorithm only accepts known plaintext for
the first part of the cipher, a future update could allow for known plaintext
to be used anywhere in the cipher.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0] - 2023-05-05

The first upload of the project.

### Added

 - evolution.py
 - main.py
 - playfair.py
 - settings.py
 - CHANGELOG.md
 - README.md
 - LICENSE
 - cipher.text
 - plain.text
 - .gitignore


main.py
from evolution import Simulation
from settings import CIPHERTEXT_FILE, MAX_GENERATIONS, PLAINTEXT_FILE, POPULATION

class Solver:

    def __init__(self, cipher_file_path: str, plain_file_path: str):
        """ Creates the Solver and loads the ciphertext and plaintext.

        Args:
            cipher_file_path: The path to the cipher.text file inc the filename.
            plain_file_path: The path to the plain.text file inc the filename.
        """
        self._load_ciphertext(cipher_file_path)
        self._load_plaintext(plain_file_path)
        self.simulation = Simulation(
                POPULATION,
                max_generations=MAX_GENERATIONS,
        )

    def solve(self):
        self.simulation.go(self.ciphertext, self.plaintext)

    def _load_ciphertext(self, cipher_file_path: str) -> None:
        """Load the ciphertext into the solver.

        This will also process the ciphertext to remove spaces and newlines.

        Args:
            cipher_file_path: The path to the cipher.text file inc the filename.
        """
        with open(cipher_file_path, 'r') as cipher_file:
            ciphertext = cipher_file.read()
            # Remove space and newline characters.
            ciphertext = ciphertext.replace('\r', '')
            ciphertext = ciphertext.replace('\n', '')
            ciphertext = ciphertext.replace(' ', '')
            if len(ciphertext) % 2:
                ciphertext = ciphertext[:-1]
            self.ciphertext = ciphertext

    def _load_plaintext(self, plain_file_path: str) -> None:
        """Load the plaintext into the solver.

        This will also process the plaintext to remove spaces and newlines.

        Args:
            cipher_file_path: The path to the cipher.text file inc the filename.
        """
        with open(plain_file_path, 'r') as plain_file:
            plaintext = plain_file.read()
            # Remove space and newline characters.
            plaintext = plaintext.replace('\r', '')
            plaintext = plaintext.replace('\n', '')
            plaintext = plaintext.replace(' ', '')

            self.plaintext = plaintext


def main():
    solver = Solver(CIPHERTEXT_FILE, PLAINTEXT_FILE)
    print(f"CIPHER TEXT: {solver.ciphertext}")
    print(f"PLAIN TEXT:  {solver.plaintext}")
    print("Beginning simulation")
    print()
    solver.solve()


if __name__ == "__main__":
    main()


playfair.py
class Playfair:

    def __init__(self, key: str):
        self.key = key
        self.key_table = {}
        x = 0
        y = 0
        for c in key:
            self.key_table[c] = (x, y)
            x = (x + 1) % 5
            if x == 0:
                y += 1

    def _char_from_index(self, x: int, y: int) -> chr:
        if x < 0 or x > 4:
            x = (x + 5) % 5
        if y < 0 or y > 4:
            y = (y + 5) % 5
        return self.key[5*y+x]

    def decrypt(self, cipher: str) -> str:
        plaintext = ""
        for i in range(0, len(cipher), 2):
            first, second = self.key_table[cipher[i]], self.key_table[cipher[i+1]] 
            if first[0] == second[0]:
                # Xs are equal, same column
                plaintext += self._char_from_index(first[0], first[1]-1)
                plaintext += self._char_from_index(second[0], second[1]-1)
            elif first[1] == second[1]:
                # Ys are equal, same row
                plaintext += self._char_from_index(first[0]-1, first[1])
                plaintext += self._char_from_index(second[0]-1, second[1])
            else:
                # Forms a rectangle
                plaintext += self._char_from_index(second[0], first[1])
                plaintext += self._char_from_index(first[0], second[1])
        return plaintext

    def __str__(self):
        return "\n".join([self.key[:5], self.key[5:10], self.key[10:15], self.key[15:20], self.key[20:]])


if __name__ == "__main__":
    # A quick test of the playfair algorithm
    playfair = Playfair("ALGORITHMSBCDEFKNPQUVWXYZ")
    cipher = "YLNREXRQATQBMLNGOVLOOIDY"
    print(f"cipher: {cipher}")
    plain = playfair.decrypt(cipher)
    print(f"plain: {plain}")
    # Best challenge solution
    playfair = Playfair("EATWHNCYSOBDFPRKLMGIVXZQU")
    cipher = "WEECGNHYLTVRRIKSAHAEEXATONIXHMNCHYWEWBXCFMHLDOSFEAIOABNEAKSRABMHKNRYBHHSDIPAHDNCHKOENCRIYRMHOYBHTWRIEABHGOSNHIDMWLHEATWYHYBNHTOPNRKNLA"
    print(len(cipher))
    print(f"cipher: {cipher}")
    plain = playfair.decrypt(cipher)
    print(f"plain: {plain}")
    # My challenge solution
    playfair = Playfair("ATWHECYSONDFPRBLMGIKXZQUV")
    cipher = "WEECGNHYLTVRRIKSAHAEEXATONIXHMNCHYWEWBXCFMHLDOSFEAIOABNEAKSRABMHKNRYBHHSDIPAHDNCHKOENCRIYRMHOYBHTWRIEABHGOSNHIDMWLHEATWYHYBNHTOPNRKNLA"
    print(len(cipher))
    print(f"cipher: {cipher}")
    plain = playfair.decrypt(cipher)
    print(f"plain: {plain}")


evolution.py
from __future__ import annotations
from random import choices, random, randint, shuffle

from playfair import Playfair
from settings import ALPHABET, KEEP_FITTEST

def new_solution_from_a(
        a: Solution,
        b: Solution,
        split: int,
        mutate: float,
    ) -> Solution:
    """Creates a new solution using A as the primary parent.
    
    The first `split` characters from A are used to create the new solution,
    the remaining characters are taken from B in order unless already selected
    from A.

    Args:
        a: The primary parent solution
        b: The secondary parent solution
        split: The index to split the key at
        mutate: The probability of a mutation
    Returns:
        Solution: A new child solution
    """
    a_key = a.key[:split]
    b_key = ""
    for c in b.key:
        if c not in a_key:
            b_key += c
    new_key = list(a_key + b_key)
    # Mutate the new key
    for i in range(len(new_key)):
        if random() < mutate:
            # Select another index, excluding itself to swap with
            swap_i = (i + randint(1, 24)) % 25
            new_key[i], new_key[swap_i] = new_key[swap_i], new_key[i]
    return Solution("".join(new_key))

def tournament_select(population: list[Solution], tournament_size: int = 2) -> Solution:
    """
    Select a solution from the given population using tournament selection
    """
    # Return the solution with the highest fitness
    solutions = choices(population, k=tournament_size)
    solutions.sort(key=lambda x: x.score)
    return solutions[-1]

class Solution(Playfair):

    def compute_score(self, cipher, target) -> None:
        count = 0
        decrypted = self.decrypt(cipher)
        for i in range(len(target)):
            count += int(decrypted[i] == target[i])
        self.score = count / len(target)

    def cross(self, other: Solution, mutate: float) -> list[Solution]:
        # Get a random split point
        split = randint(1, 23)
        return [
            new_solution_from_a(self, other, split, mutate),
            new_solution_from_a(other, self, split, mutate),
        ]

def random_key():
    alphabet = list(ALPHABET)
    shuffle(alphabet)
    return "".join(alphabet)

class Simulation:

    def __init__(self, population_size: int, mutate: float = 0.1, max_generations: int = 1000):
        self.max_generations = max_generations
        self.population_size = population_size
        self.mutate = mutate
        self.population = [
                Solution(random_key())
                for i in range(population_size)
        ]

    def go(self, cipher_text, plain_text):
        generation = 0
        while generation < self.max_generations:
            generation += 1
            for s in self.population:
                s.compute_score(cipher_text, plain_text)
            self.population.sort(key=lambda x: x.score)
            if self.population[-1].score == 1:
                print()
                print("SOLUTION FOUND!")
                print(str(self.population[-1]))
                print(self.population[-1].decrypt(cipher_text))
                return
            print(f"Generation: {generation}, Fitness: {self.population[-1].score:.4f}", end='\r')
            if KEEP_FITTEST:
                new_population = self.population[-KEEP_FITTEST:]
            for i in range((self.population_size // 2) - KEEP_FITTEST):
                parent_a = tournament_select(self.population)
                parent_b = tournament_select(self.population)
                children = parent_a.cross(parent_b, self.mutate)
                new_population += children
            self.population = new_population
        for s in self.population:
            s.compute_score(cipher_text, plain_text)
        self.population.sort(key=lambda x: x.score)
        print()
        print("Best:")
        print(str(self.population[-1]))
        print(self.population[-1].decrypt(cipher_text))



settings.py
"""Settings"""

# Alphabet is the letters that will be used to build the table
ALPHABET = "ABCDEFGHIKLMNOPQRSTUVWXYZ"  # Exclude J from the alphabet

# The maximum number of generations until the program stops and outputs the
# current best answer
MAX_GENERATIONS = 10000

# The number of solutions in each generation
# This should be an even number and greater than KEEP_FITTEST
POPULATION = 1500

# The number of best solutions carried over from the previous generation
# This should be an even number
KEEP_FITTEST = 30

# The path to the cipher text file
CIPHERTEXT_FILE = "cipher.text"

# The path to the plain text file (The known plain text of the cipher)
PLAINTEXT_FILE = "plain.text"

def validate_settings():
    errors = []
    warnings = []

    # Validate Alphabet
    if not len(ALPHABET) == 25:
        errors.append("The alphabet must contain 25 characters")
    if not len(set(ALPHABET)) == len(ALPHABET):
        errors.append("The alphabet must contain unique characters")
    if not ALPHABET == ALPHABET.upper():
        warnings.append("Letters in the alphabet should be uppercase")

    # Validate Max Generations
    if MAX_GENERATIONS < 1:
        errors.append("The Max Generations must be a positive integer")
    if MAX_GENERATIONS < 15:
        warnings.append("The Max Generations is too low, you may not get good "
                        "results")
    
    # Validate Population
    if POPULATION < 1:
        errors.append("The Population size must be a positive integer")
    if POPULATION < 10:
        warnings.append("The Population size is small, you may not get good "
                        "results")
    if POPULATION - KEEP_FITTEST < 1:
        errors.append("The Population must be larger than KEEP_FITTEST")

    # Validate Keep Fittest
    if KEEP_FITTEST < 0:
        errors.append("Keep Fittest must be 0 or a positive integer")
    if POPULATION - KEEP_FITTEST < POPULATION // 2:
        warnings.append("More than half the population is kept between "
                        "generations, you may not get good results")

    if warnings:
        for warning in warnings:
            print(f"WARNING: {warning}")
    if errors:
        for error in errors:
            print(f"ERROR: {error}")
        exit()

    # We don't bother validating the ciphertext and plaintext files as they
    # will error at the start of the program.

validate_settings()

